### DISCUSSION TITLE
v0.17.0: bitwise ops, triplet losses, new layers, fixes, and more.

### DISCUSSION THREAD

**Discussion by janpfeifer** [gomlx/gomlx#117] [Announcements]
* Backend API change: separating Logical and Bitwise versions of various ops derived from And, Or, Xor and Not.
* Updated dependency to gopjrt v0.6.0.
* Added "Flow Matching" examples/demo.
* Package `layers`:
  * Added `layers.DropBlock`, a type of dropout for images.
  * Added `layers.DropPath` and `layers.DropPathFromContext`, a type of dropout used in Residual connections, to drop full paths.
  * `layers.LayerNormalization`:
    * up-scale precision by default if input is a Float16 or BFloat16. Low-precision
      lead to NaNs when reducing values for normalization. Added also a hyperparameter to configure normalization DType.
* Added `Context.RandomBenoulli` to sample from a Bernoulli (binary) distribution.
* Correctly pretty-print Float16 and BFloat16 tensors.
* Several fixes and small improvements to command-line tool `gomlx_checkpoint`.
* Package `nanlogger`:
  * Store only the stack-trace, and trim the stack into the nanlogger package.
  * Does not exit, simply report the NanLogger. User can define a handler, if they want the training to exit.
  * Use `IsFinite` to check for NaN and Infs: but we loose the type of NaN that happened.  
  * Fixed nanlogger for Float16 and BFloat16; Also, it first prints other logged tensors, before failing with a NaN.
* Package `losses`: 
  * Added `ParamLoss`: hyperparameter to define the loss, and many constant values.
  * Added `LossFromContext`, using `ParamLoss` hyperparameter. 
  * Added `MakeHuberLossFromContext`
  * Added experimental `MakeAdaptivePowerLoss` and `MakeAdaptivePowerLossFromContext`
  * Added TripletLoss: various negative sampling strategies and distance metrics.
* Package `graph`:
  * More unit tests.
  * Aliases nodes: allow setting aliases to nodes, and to retrieve them by those aliases. Useful for layers
    or models to export intermediary nodes by their aliases. They are prefixed by scope. New methods are:
    `Node.WithAlias`, `Node.GetAlias`, `Graph.GetNodeByAlias`, `Graph.PushAliasScope`, `Graph.PopAliasScope` 
    and `Graph.IterAliasedNodes`.
  * Added optional aliases nodes for `inceptionv3` model.
  * Added `ReduceSkewness` and the alias `Skewness`.
  * Added bitwise ops:
    * `BitwiseShiftLeft`, `BitwiseShiftRightLogical`, `BitwiseShiftRightArithmetic`, `BitwiseAnd`, `BitwiseOr`,
      `BitwiseXor`, `BitwiseNot`.
  * Kept an alias from `And`, `Or` and `Not` to the `LogicalAnd`, `LogicalOr`, `LogicalXor` and `LogicalNot`.

<hr /><em>This discussion was created from the release <a href='https://github.com/gomlx/gomlx/releases/tag/v0.17.0'>v0.17.0: bitwise ops, triplet losses, new layers, fixes, and more.</a>.</em>
